consensus/core/src/core.rs

```diff
@@ -47,37 +50,55 @@ impl Core {
         block_manager: BlockManager,
         mut signals: CoreSignals,
         block_signer: NetworkKeyPair,
+        store: Arc<dyn Store>,
     ) -> Self {
-        // TODO: restore the threshold clock round based on the last quorum data in storage when crash/recover
-        let mut threshold_clock = ThresholdClock::new(0, context.clone());
+        // Recover the pending ancestors to ensure that we'll have a quorum of blocks to make progress.
+        let (mut last_proposed_block, mut all_pending_ancestors) = Block::genesis(context.clone());
 
-        // TODO: restore based on DagState, for now we just init via the genesis
-        let (genesis_my, mut genesis_others) = Block::genesis(context.clone());
-        genesis_others.push(genesis_my.clone());
+        // Now fetch the proposed blocks per authority for their last two rounds.
+        for (index, _authority) in context.committee.authorities() {
+            let blocks = store
+                .scan_last_blocks_by_author(index, 2)
+                .expect("Storage error while recovering Core");
 
-        // populate the threshold clock to properly advance the round & also the pending ancestors
-        let mut pending_ancestors: BTreeMap<Round, Vec<VerifiedBlock>> = BTreeMap::new();
-        for ancestor in genesis_others {
-            threshold_clock.add_block(ancestor.reference());
-            pending_ancestors
-                .entry(ancestor.round())
-                .or_default()
-                .push(ancestor)
+            if index == context.own_index && !blocks.is_empty() {
+                last_proposed_block = blocks.last().cloned().unwrap();
+            }
+            all_pending_ancestors.extend(blocks);
         }
 
+        // populate the threshold clock to properly advance the round & also populate the pending ancestors
+        let mut threshold_clock = ThresholdClock::new(0, context.clone());
```

Comment:
Leave a comment on why no block will be proposed at or below the round of `last_proposed_block`, after recovery?